generator/restriction_collector.cpp

```diff
@@ -0,0 +1,218 @@
+#include "std/sstream.hpp"
+
+namespace
+{
+char const kNo[] = "No";
+char const kOnly[] = "Only";
+
+bool ParseLineOfNumbers(istringstream & stream, vector<uint64_t> & numbers)
+{
+  string numberStr;
+  uint64_t number;
+
+  while (stream)
+  {
+    if (!getline(stream, numberStr, ','))
+      return true;
+    if (numberStr.empty())
+      return true;
+
+    if (!strings::to_uint64(numberStr, number))
+      return false;
+
+    numbers.push_back(number);
+  }
+  return true;
+}
+}  // namespace
+
+namespace routing
+{
+RestrictionCollector::RestrictionCollector(string const & restrictionPath,
+                                           string const & featureId2OsmIdsPath)
+{
+  if (!ParseFeatureId2OsmIdsMapping(featureId2OsmIdsPath))
+    return;
+
+  if (!ParseRestrictions(restrictionPath))
+  {
+    m_restrictions.clear();
+    return;
+  }
+  ComposeRestrictions();
+  RemoveInvalidRestrictions();
+  LOG(LINFO, ("m_restrictions.size() =", m_restrictions.size()));
+}
+
+bool RestrictionCollector::IsValid() const
+{
+  return !m_restrictions.empty() &&
+         find_if(begin(m_restrictions), end(m_restrictions),
+                 [](Restriction const & r) { return !r.IsValid(); }) == end(m_restrictions);
+}
+
+bool RestrictionCollector::ParseFeatureId2OsmIdsMapping(string const & featureId2OsmIdsPath)
+{
+  ifstream featureId2OsmIdsStream(featureId2OsmIdsPath);
+  if (featureId2OsmIdsStream.fail())
+    return false;
+
+  while (featureId2OsmIdsStream)
+  {
+    string line;
+    if (!getline(featureId2OsmIdsStream, line))
+      return true;
+
+    istringstream lineStream(line);
+    vector<uint64_t> ids;
+    if (!ParseLineOfNumbers(lineStream, ids))
+      return false;
+
+    if (ids.size() <= 1)
+      return false;  // Every line should contain at least feature id and osm id.
+
+    Restriction::FeatureId const featureId = static_cast<Restriction::FeatureId>(ids.front());
+    ids.erase(ids.begin());
+    AddFeatureId(featureId, ids);
+  }
+  return true;
+}
+
+bool RestrictionCollector::ParseRestrictions(string const & restrictionPath)
+{
+  ifstream restrictionsStream(restrictionPath);
+  if (restrictionsStream.fail())
+    return false;
+
+  while (restrictionsStream)
+  {
+    string line;
+    if (!getline(restrictionsStream, line))
+      return true;
+    istringstream lineStream(line);
+    string typeStr;
+    getline(lineStream, typeStr, ',');
+    Restriction::Type type;
+    if (!FromString(typeStr, type))
+      return false;
+
+    vector<uint64_t> osmIds;
+    if (!ParseLineOfNumbers(lineStream, osmIds))
+      return false;
+
+    AddRestriction(type, osmIds);
+  }
+  return true;
+}
+
+void RestrictionCollector::ComposeRestrictions()
+{
+  // Going through all osm id saved in |m_restrictionIndex| (mentioned in restrictions).
+  size_t const restrictionSz = m_restrictions.size();
+  for (pair<uint64_t, Index> const & osmIdAndIndex : m_restrictionIndex)
+  {
+    Index const & index = osmIdAndIndex.second;
```

Comment:
We already have Index in our project, moreover, this class is quite important and is used almost everywhere. So, this name is not good.